### PR TITLE
Add a Github preview webhook

### PR DESCRIPTION
**Pull request by samuelli**
This adds a Github preview webhook which requires authorization. Once authorization is granted, a webhook is configured on the specified repository. All pull-requests will then receive a status check with a link to the processed documentation on custom-elements.io.

### CODE REVIEW

File: src/manage.py [hunk 1]
```diff
@@ -419,6 +438,7 @@ def get(self):
     webapp2.Route(r'/manage/delete_everything/yes_i_know_what_i_am_doing', handler=DeleteEverything),
     webapp2.Route(r'/task/update/<owner>/<repo>', handler=UpdateLibrary),
     webapp2.Route(r'/task/ingest/commit/<owner>/<repo>', handler=IngestLibraryCommit),
+    webapp2.Route(r'/task/ingest/webhook/<owner>/<repo>', handler=IngestWebhookLibrary),
```
Comment (dstockwell) [inline src/manage.py]: add this url to the xsrf test in manage_test

Comment (samuelli) [inline src/manage.py]: Done.

**Comment by samuelli** [comment]
ptal

File: src/api.py [hunk 2]
```diff
@@ -187,24 +187,126 @@ def get(self, owner, repo, ver=None):
...
+    # Extract Github app details
+    client_id = util.SECRETS['github_client_id']
+    client_secret = util.SECRETS['github_client_secret']
 
-    response = urlfetch.fetch('https://github.com/login/oauth/access_token?client_id=%s&client_secret=%s&code=%s' %
-                              (client_id, client_secret, code), method='POST', validate_certificate=True)
+    # Exchange code for an access token from Github
+    headers = {'Accept': 'application/json'}
+    access_token_url = 'https://github.com/login/oauth/access_token?client_id=%s&client_secret=%s&code=%s' % (client_id, client_secret, code)
+    access_response = urlfetch.fetch(access_token_url, headers=headers, method='POST', validate_certificate=True)
```
Comment (dstockwell) [inline src/api.py]: Can we pass the params in the payload?

File: src/api.py [hunk 3]
```diff
@@ -187,24 +187,126 @@ def get(self, owner, repo, ver=None):
...
+    access_token_response = json.loads(access_response.content)
```
Comment (dstockwell) [inline src/api.py]: Check the response code.

File: src/api.py [hunk 4]
```diff
@@ -187,24 +187,126 @@ def get(self, owner, repo, ver=None):
...
+      self.response.set_status(401)
+      self.response.write('Authorization failed')
+      return
+    access_token = access_token_response['access_token']
+
+    # Validate access token against repo
+    repos_response = util.github_resource('user/repos', access_token=access_token)
+    if repos_response.status_code != 200:
+      self.response.set_status(401)
+      self.response.write('Cannot access users repos')
```
Comment (dstockwell) [inline src/api.py]: `user's`

File: src/api.py [hunk 5]
```diff
@@ -187,24 +187,126 @@ def get(self, owner, repo, ver=None):
...
+    # Check if the webhook exists
+    list_webhooks_response = util.github_post('repos', owner, repo, 'hooks', access_token=access_token)
+    if list_webhooks_response.status_code != 200:
+      logging.error('Unable to query existing webhooks, continuing anyway. Github %s: %s',
+                    list_webhooks_response.status_code, list_webhooks_response.content)
+    else:
+      webhooks = json.loads(list_webhooks_response.content)
+      for webhook in webhooks:
+        if webhook['active'] and webhook['config'] == params['config']:
+          self.response.set_status(202)
```
Comment (dstockwell) [inline src/api.py]: Should this just be `200`, does the client really care?

File: src/datamodel.py [hunk 6]
```diff
@@ -9,6 +9,8 @@ class CollectionReference(ndb.Model):
 class Library(ndb.Model):
   kind = ndb.StringProperty()
 
+  access_token = ndb.StringProperty()
```
Comment (dstockwell) [inline src/datamodel.py]: Perhaps a more descriptive name, `github_status_access_token` or something.

**Comment by dstockwell** [comment]
lgtm

File: src/api.py (hunk 2 above)
Comment (samuelli) [inline src/api.py]: Done.

File: src/api.py (hunk 3 above)
Comment (samuelli) [inline src/api.py]: Done

File: src/api.py (hunk 4 above)
Comment (samuelli) [inline src/api.py]: Done

File: src/api.py (hunk 5 above)
Comment (samuelli) [inline src/api.py]: Done

**Pull request merged**